BJS2-94449 Solution of the task

KuaNySH88 opened the pull request.

KuaNySH88 committed:
BJS2-94449 Solution of the task

MarselYsup commented on src/main/java/school/faang/bjs2_94449/QuestSystem.java:
```diff
@@ -0,0 +1,21 @@
+            throws ExecutionException, InterruptedException {
+        CompletableFuture<Player> playerCompletableFuture = CompletableFuture.supplyAsync(() -> {
+            try {
+                Thread.sleep((long) quest.difficulty() * 100);
```
100 - магическое число, давай вынесем в константу

KuaNySH88 replied:
поправил

MarselYsup commented on src/main/java/school/faang/bjs2_94449/QuestSystem.java:
```diff
@@ -0,0 +1,21 @@
+            }
+            player.addExperience(quest.reward());
+            return player;
+        });
```
Использование пула потоков по умолчанию (ForkJoinPool.commonPool) не рекомендуется в производственном коде, так как вы не контролируете его поведение. Передавайте свой собственный пул потоков в качестве аргумента.
```java
CompletableFuture.supplyAsync(() -> {...}, executorService);
```

Подставляй вторым аргументом свой executorService

KuaNySH88 replied:
добавил

MarselYsup commented on src/main/java/school/faang/bjs2_94449/QuestSystem.java:
```diff
@@ -0,0 +1,21 @@
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            player.addExperience(quest.reward());
```
Если один игрок выполняет несколько квестов одновременно, может возникнуть состояние гонки. Добавь блокировки (synchronized или другие способы блокировки) или используй потокобезопасные структуры. например AtomicInteger

KuaNySH88 replied:
поправил

MarselYsup commented on src/main/java/school/faang/bjs2_94449/Player.java:
```diff
@@ -0,0 +1,16 @@
+@AllArgsConstructor
+@Getter
+public class Player {
+    String name;
```
используй модификатор доступа private для внутренних полей класса , следуй принципу инкапсуляции

KuaNySH88 replied:
поправил

KuaNySH88 committed:
BJS2-94449 Fix PR comments:
1. add own executorService in CompletableFuture
2. make private the variable Player.name
3. make synchronized method Player.addExperience
4. create constant for magic number
